### PR TITLE
randomsub implementation

### PR DESCRIPTION
**Pull request by vyzo**
So that we have the whole zoo; people actually want to use it, even if it's just for testing.

### CODE REVIEW

**Comment by bigs** [comment]
u crazy for this one

File: randomsub.go [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+				tosend[p] = struct{}{}
+			} else {
+				rspeers[p] = struct{}{}
+			}
+		}
+	}
+
+	if len(rspeers) > RandomSubD {
+		xpeers := peerMapToList(rspeers)
+		shufflePeers(xpeers)
```
Comment (bigs) [inline randomsub.go:80]: unnecessary, as in golang map access is randomized!

Comment (bigs) [reply randomsub.go:80]: we can just count how many we've iterated through and break when we've hit our limit

**Review by bigs — CHANGES_REQUESTED** [review]

File: randomsub.go (hunk 1 above)
Comment (vyzo) [reply randomsub.go:80]: afaik it is simply non-deterministic, not randomized between iterations.

**Review by bigs — APPROVED** [review]

File: randomsub.go (hunk 1 above)
Comment (vyzo) [reply randomsub.go:80]: more precisely, it simply starts from a different (random) point in the map on each iteration -- the relative order of the keys will not change if there haven't been any insertions in the map.

**Comment by raulk** [comment]
Would be good to get a test for this at some point.

**Comment by raulk** [comment]
(not intending to be the party pooper 😛)

**Comment by vyzo** [comment]
feel free to write one :)

**Comment by raulk** [comment]
@vyzo here it is: https://github.com/raulk/go-floodsub/commit/8ee2d829d720d322360960a46be665ac01b31bc6

Could you please check out why it fails with randomsub but succeeds with floodsub and gossipsub?

Perhaps randomsub doesn't have the expected emergent properties, but maybe there's something wrong with the test!

EDIT: Let me know when you want it in the codebase, and I'll send in a PR.

**Comment by vyzo** [comment]
@raulk It's not unexpected that it would lose some messages for some peers -- there is a non-zero probability that the random propagation will not deliver a message to a node. That's why we don't use it in production (and the main reason for why I didn't bother with the test in the first place).

**Pull request merged**